core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBasedDeterministicHashPolicy.java

```diff
@@ -0,0 +1,135 @@
+
+/**
+ * A policy that pseudo-randomly distributes blocks between workers according to their capacity,
+ * so that the probability a worker is chosen is equal to the ratio of its capacity over total
+ * capacity of all workers, provided that the blocks requested follow a uniform distribution.
+ * If sharding is disabled, the same block is always assigned to the same worker. If sharding
+ * is enabled, the block is assigned to a fixed set of workers.
+ *
+ * The difference between this policy and {@link CapacityBaseRandomPolicy} is that this policy
+ * uses the hashed block ID as the index to choose the target worker, so that the same block is
+ * always routed to the same set of workers.
+ *
+ * Both this policy and {@link DeterministicHashPolicy} choose workers based the hashed block ID.
+ * The difference is that {@link DeterministicHashPolicy} uniformly distributes the blocks among
+ * the configured number of shards, while this policy chooses workers based on a distribution of
+ * their normalized capacity.
+ *
+ * @see CapacityBaseRandomPolicy
+ * @see DeterministicHashPolicy
+ */
+public class CapacityBasedDeterministicHashPolicy implements BlockLocationPolicy {
+  private final int mShards;
+
+  /**
+   * Constructor required by
+   * {@link BlockLocationPolicy.Factory#create(Class, AlluxioConfiguration)}.
+   * @param conf Alluxio configuration
+   */
+  public CapacityBasedDeterministicHashPolicy(AlluxioConfiguration conf) {
+    int numShards =
+        conf.getInt(PropertyKey.USER_UFS_BLOCK_READ_LOCATION_POLICY_DETERMINISTIC_HASH_SHARDS);
+    Preconditions.checkArgument(numShards >= 1, "number of shards must be no less than 1");
+    mShards = numShards;
+  }
+
+  @Override
+  public Optional<WorkerNetAddress> getWorker(GetWorkerOptions options) {
+    // the target worker is determined by the following algorithm:
+    // 1. build a cumulative distribution function by adding up all workers and their capacities.
+    //    workers are sorted by their host name alphabetically.
+    //    if worker A has 90 GB, B has 10 GB and C has 900 GB, the CDF looks like
+    //    | 0 ... 90 | 90 ... 100 | 100 ... 1000 |
+    //    | worker A |  worker B  |   worker C   |
+    // 2. find a fixed starting point in [0, totalCapacity) determined by the hashed block id.
+    //    | 0 ... 90 | 90 ... 100 | 100 ... 1000 |
+    //    | worker A |  worker B  |   worker C   |
+    //                     ^ start = 95
+    // 3. find the corresponding worker in the CDF.
+    //    which is worker B in this example
+    // 4. if #shards = 1, this worker is selected. otherwise, find a set of candidates:
+    //    4.1 hashed_block_id(0) = block id
+    //    4.2 for i in [1, #shards], hashed_block_id(i) = hash(hashed_block_id(i-1))
+    //    4.3 find the worker whose position corresponds to hashed_block_id(i) in the CDF,
+    //        and add it to the candidates set
+    //    4.4 repeat 4.2 - 4.4
+    // 5. select a random worker in the candidate set
```

Comment:
nit: put the these algorithm description to the class level java doc?  your call

Reply:
Done